// peer.io/lib/user.js
/*
 * User represents a peer together with a list of other peers
 * he is attempting to connect with.
 * 
 * Peer interest is added when a peer makes a connection
 * request. Interest is removed after 5 minutes. 
 * 
 * @author Daniel Seah
 */
function User(id, sock) {
	this.id = id;
	var interests = {};
	this.socket = sock;
	
	this.addInterest = function(pid, label) {
		if(typeof interests[pid] === 'undefined') {
			interests[pid] = {};
		}
		interests[pid][label] = new Date();
	};
	
	this.isInterested = function(pid, label) {
		if(typeof interests[pid] === 'undefined') {
			return false;
		}
		if(typeof interests[pid][label] === 'undefined') {
			return false;
		}
		var now = new Date().getTime();
		var elapsed = now - interests[pid][label].getTime();
		if(elapsed > (1000*60*5)) { // 5 minutes
			return false;
		}
		return true;
	};
	
	this.removeInterest = function(pid, label) {
		if(typeof interests[pid] === 'undefined') {
			return;
		}
		if(typeof interests[pid][label] === 'undefined') {
			return;
		}
		delete interests[pid][label];
	};
}

module.exports = User;
